---
title: "R Notebook"
output: html_notebook
---

```{r}
library(knitr)
opts_chunk$set(message=FALSE, warning=FALSE, eval=TRUE, echo=TRUE, cache=TRUE)
.ex <- 1
library(ggplot2)
theme_set(theme_bw(base_size=16) + theme(strip.background = element_blank()))
library(readr)
devtools::install_github('shenorrLab/FIT.mouse2man')
library(dplyr)
library(plyr)

```

use biomart to find the orthologs annotated between mouse and human, and translate all ensemble ids into entrez ids

```{r}
human_ensemble2entrez <- read_csv("C:/Users/inbal/Google Drive/shen-orr/Biomart/Biomart_human_ensemble2entrez.csv")
mouse_ensemble2entrez <- read_csv("C:/Users/inbal/Google Drive/shen-orr/Biomart/Biomart_mouse_ensemble2entrez.csv")
mouse_human_orthologs <- read_csv("C:/Users/inbal/Google Drive/shen-orr/Biomart/Biomart_with_human_ortholog.csv")
entrez_orthologs <- merge(x=mouse_ensemble2entrez,y=mouse_human_orthologs,by.x="mouse_Gene_stable_ID",by.y="mouse_Gene_stable_ID")
entrez_orthologs <- merge(x=entrez_orthologs,y=human_ensemble2entrez, by.x = "Human_gene_stable_ID", by.y = "human_Gene_stable_ID")
#NCBI_gene_ID = entrez ID, Gene_stable_ID = ensemble ID
entrez_orthologs
```

Biomart: find (1 -> many), (many->many) genes
```{r}
mouse_genes = unique(entrez_orthologs$mouse_NCBI_gene_ID)
mouse_genes = mouse_genes[!is.na(mouse_genes)]
one2many_res<-c()
many2many_res<-c()
for (mouse_gene in mouse_genes) {
  one2many=TRUE
  many2many=TRUE
  human_orthologs <- entrez_orthologs %>% filter(mouse_NCBI_gene_ID==mouse_gene) #find how many different human orthologs this mouse gene has.
  human_orthologs <- human_orthologs[!is.na(human_orthologs$human_NCBI_gene_ID),] #remove missing values in human entrez id.
  if (length(unique(human_orthologs$human_NCBI_gene_ID)) <=1) {
    one2many=FALSE
    many2many=FALSE
  }
  if (length(unique(human_orthologs$human_NCBI_gene_ID)) >1) {
    for(human_gene in unique(human_orthologs$human_NCBI_gene_ID)){
      if (entrez_orthologs %>% filter(human_NCBI_gene_ID==human_gene) %>% nrow() >1) one2many=FALSE
    }
  }
  if (one2many==TRUE) {
    one2many_res<-c(one2many_res,mouse_gene)
  }
  else if (many2many==TRUE){
    many2many_res<-c(many2many_res,mouse_gene)
  }
}
write(many2many_res, "C:/Users/inbal/Google Drive/shen-orr/Biomart/many2many_entrezID.txt", ncolumns = 1)
write(one2many_res, "C:/Users/inbal/Google Drive/shen-orr/Biomart/one2many_entrezID.txt", ncolumns = 1)

```
Biomart with types

```{r}
human_ensemble2entrez <- read_csv("C:/Users/inbal/Google Drive/shen-orr/Biomart/Biomart_human_ensemble2entrez.csv")
mouse_ensemble2entrez <- read_csv("C:/Users/inbal/Google Drive/shen-orr/Biomart/Biomart_mouse_ensemble2entrez.csv")
mouse_human_orthologs <- read_csv("C:/Users/inbal/Google Drive/shen-orr/Biomart/Biomart_orthologs_withtypes.csv")
entrez_orthologs <- merge(x=mouse_ensemble2entrez,y=mouse_human_orthologs,by.x="mouse_Gene_stable_ID",by.y="mouse_Gene_stable_ID")
entrez_orthologs <- merge(x=entrez_orthologs,y=human_ensemble2entrez, by.x = "Human_gene_stable_ID", by.y = "human_Gene_stable_ID")
#NCBI_gene_ID = entrez ID, Gene_stable_ID = ensemble ID
entrez_orthologs <- select(entrez_orthologs, c('mouse_NCBI_gene_ID','Human_homology_type','human_NCBI_gene_ID'))
entrez_orthologs <- entrez_orthologs[!(is.na(entrez_orthologs$human_NCBI_gene_ID)),]
entrez_orthologs <- entrez_orthologs[!(is.na(entrez_orthologs$mouse_NCBI_gene_ID)),]

Biomart_one2one = unique((entrez_orthologs %>% filter(Human_homology_type=='ortholog_one2one'))$mouse_NCBI_gene_ID)
Biomart_one2many = unique((entrez_orthologs %>% filter(Human_homology_type=='ortholog_one2many'))$mouse_NCBI_gene_ID)
Biomart_many2many = unique((entrez_orthologs %>% filter(Human_homology_type=='ortholog_many2many'))$mouse_NCBI_gene_ID)

cat(length(Biomart_one2one),'one2one orthologs\n' )
cat(length(Biomart_one2many),'one2many orthologs\n' )
cat(length(Biomart_many2many),'many2many orthologs\n' )

```

OMA database

```{r}
OMA_orthologs <- read_delim("C:/Users/inbal/Google Drive/shen-orr/OMA/OMA_human_mouse_orthologs.csv", delim = '\t',col_names = c('human_entrezID', 'mouse_entrezID','ortholog_type','OMA_group'))
OMA_orthologs

OMA_one2one = unique((OMA_orthologs %>% filter(ortholog_type=='1:1'))$mouse_entrezID)
OMA_one2many = unique((OMA_orthologs %>% filter(ortholog_type=='1:m'))$mouse_entrezID)
OMA_many2many = unique((OMA_orthologs %>% filter(ortholog_type=='n:m'))$mouse_entrezID)
OMA_many2one = unique((OMA_orthologs %>% filter(ortholog_type=='n:1'))$mouse_entrezID)

cat(length(Biomart_one2one),'Biomart one2one orthologs\n' )
cat(length(Biomart_one2many),'Biomart one2many orthologs\n' )
cat(length(Biomart_many2many),'Biomart many2many orthologs\n\n\n' )


cat(length(OMA_one2one),'OMA one2one orthologs\n' )
cat(length(OMA_one2many),'OMA one2many orthologs\n' )
cat(length(OMA_many2many),'OMA many2many orthologs\n' )
cat(length(OMA_many2one),'OMA many2one orthologs\n\n\n' )

cat(length(intersect(OMA_one2one,Biomart_one2one)),'one2one orthologs are common to biomart and OMA\n' )
cat(length(intersect(OMA_one2many,Biomart_one2many)),'one2many orthologs are common to biomart and OMA\n' )
cat(length(intersect(OMA_many2many,Biomart_many2many)),'many2many orthologs are common to biomart and OMA\n' )

```


MGI

```{r}
# download from http://www.informatics.jax.org/downloads/reports/index.html#homology

human_mouse_orthologs <- read_delim("C:/Users/inbal/Google Drive/shen-orr/MGI/HOM_MouseHumanSequence.csv",delim='\t')

mouse_entrezID=c()
human_entrezID=c()

for (ortholog_num in unique(human_mouse_orthologs$HomoloGene_ID)){
  MouseIdTemp = human_mouse_orthologs %>% filter(HomoloGene_ID==ortholog_num,Common_Organism_Name=="mouse, laboratory")
  MouseIdTemp = MouseIdTemp$EntrezGene_ID
  HumanIdTemp = human_mouse_orthologs %>% filter(HomoloGene_ID==ortholog_num,Common_Organism_Name=="human")
  HumanIdTemp = HumanIdTemp$EntrezGene_ID
  tempdf = expand.grid(HumanIdTemp,MouseIdTemp)
  mouse_entrezID = append(mouse_entrezID, tempdf$Var2)
  human_entrezID = append(human_entrezID,tempdf$Var1)
}


human_mouse_orthologs = data.frame(mouse_entrezID, human_entrezID)
human_mouse_orthologs

```
MGI: find (1 -> many), (many->many) genes
```{r}
mouse_genes = unique(human_mouse_orthologs$mouse_entrezID)
one2many_MGI<-c()
many2many_MGI<-c()
for (mouse_gene in mouse_genes) {
  one2many=TRUE
  many2many=TRUE
  human_orthologs <- human_mouse_orthologs %>% filter(mouse_entrezID==mouse_gene) #find how many different human orthologs this mouse gene has.
  if (length(unique(human_orthologs$human_entrezID)) <=1) {
    one2many=FALSE
    many2many=FALSE
  }
  if (length(unique(human_orthologs$human_entrezID)) >1) {
    for(human_gene in unique(human_orthologs$human_entrezID)){
      if (human_mouse_orthologs %>% filter(human_entrezID==human_gene) %>% nrow() >1) one2many=FALSE
    }
  }
  if (one2many==TRUE) {
    one2many_MGI <- c(one2many_MGI,mouse_gene)
  }
  else if (many2many == TRUE){
    many2many_MGI <- c(many2many_MGI,mouse_gene)
  }
}

write(many2many_MGI, "C:/Users/inbal/Google Drive/shen-orr/MGI/many2many_entrezID.txt", ncolumns = 1)
write(one2many_MGI, "C:/Users/inbal/Google Drive/shen-orr/MGI/one2many_entrezID.txt", ncolumns = 1)

```


#compare the 2 datasets:
```{r}

cat('in the Biomart results i found:',length(one2many_res),"one -> many genes.\n")
cat(length(many2many_res),"many -> many genes.\n\n")

cat("in the MGI results i found:",length(one2many_MGI),"one -> many genes.\n")
cat(length(many2many_MGI),"many -> many genes.\n\n")

cat(length(intersect(one2many_res, one2many_MGI)),"(one -> many) genes were found in both\n")

cat(length(intersect(many2many_res, many2many_MGI)),"(many -> many) genes were found in both\n")
```

explore the fit data.

```{r}
data("AllData_V2.0",package="FIT.mouse2man")

cat(length(intersect(unique(AllData_V2.0$MM.Entrez),one2many_res)),"(out of 250) of the one to many genes found by biomart are considered one to one in the FIT model\n\n")

cat(length(intersect(unique(AllData_V2.0$MM.Entrez),many2many_res)),"(out of 603) of the many to many genes found by biomart are considered one to one in the FIT model\n")


```


wormhole

```{r}
isInt <- function(N){
    !grepl("[^[:digit:]]", format(N,  digits = 20, scientific = FALSE))
}
convertID2EntrezID <- function(ID){
    return((wormholeID_to_entrez %>% filter(Input==ID)) $ "Entrez Gene ID")
}

wormholeID_to_entrez <- select(read_delim("C:/Users/inbal/Google Drive/shen-orr/WORMHOLE/wormholeID_to_entrez.txt",delim='\t',col_types = "ccci", col_names = c("wormholeID","_","_1","Entrez_gene_ID"),skip = 1),c("wormholeID","Entrez_gene_ID"))
wormhole_orthologs <- select(read_delim("C:/Users/inbal/Google Drive/shen-orr/WORMHOLE/hsmm-WORMHOLE-orthologs.txt",delim='\t'),c("Query.ID","Target.ID","Votes","Vote.Score","WORMHOLE.Score"))
wormhole_orthologs <- wormhole_orthologs %>% rename(
    human.ID = Query.ID,
    mouse.ID = Target.ID
    )
wormhole_orthologs_temp <- select(read_delim("C:/Users/inbal/Google Drive/shen-orr/WORMHOLE/mmhs-WORMHOLE-orthologs.txt",delim='\t'),c("Query.ID","Target.ID","Votes","Vote.Score","WORMHOLE.Score"))
wormhole_orthologs_temp <- wormhole_orthologs_temp %>% rename(
    mouse.ID = Query.ID,
    human.ID = Target.ID
    )
wormhole_orthologs <- rbind(wormhole_orthologs,wormhole_orthologs_temp)
rm(wormhole_orthologs_temp)

temp <- wormhole_orthologs %>% filter(Votes > 1)
temp <- merge(temp, wormholeID_to_entrez, by.x = ,by.y = )


```

panther

```{r}

#read in file
panther_orthologs <- read_delim("C:/Users/inbal/Google Drive/shen-orr/panther/QfO_Genome_Orthologs_pairwise",delim='\t', col_names = c("gene1","gene2","ortholog_type","_","common_ancestor"),col_types = "ccccc")

# filter only the mouse- human orthologs
panther_orthologs <- panther_orthologs %>% filter(((startsWith(gene1,"MOUSE") | startsWith(gene1,"HUMAN")) & (startsWith(gene2,"MOUSE") | startsWith(gene2,"HUMAN"))))

#parse out the ids and id types
panther_orthologs["human_geneID_type"]=sapply(strsplit(panther_orthologs$gene1,"\\||="),"[",2)
panther_orthologs["human_gene_id"]=sapply(strsplit(panther_orthologs$gene1,"\\||="),"[",3)

panther_orthologs["mouse_geneID_type"]=sapply(strsplit(panther_orthologs$gene2,"\\||="),"[",2)
panther_orthologs["mouse_gene_id"]=sapply(strsplit(panther_orthologs$gene2,"\\||="),"[",4)

#filter to only use MGI/ hgnc IDs bc the count is:
#mouse:Ensembl 56, Gene 423, GeneID 1, MGI 25007
#human:Ensembl 465, Gene 73, Gene_ORFName 4, GeneCards 1, GeneID 2, HGNC 24942
panther_orthologs <- select(panther_orthologs, c("ortholog_type","human_geneID_type","human_gene_id","mouse_geneID_type","mouse_gene_id"))
panther_orthologs <- panther_orthologs %>% filter(human_geneID_type=="HGNC" & mouse_geneID_type=="MGI")

#convert IDS to entrez
hgnc2entrez <- read_delim("C:/Users/inbal/Google Drive/shen-orr/panther/hgnc2entrez.txt", delim = "\t", col_names = TRUE)
hgnc2entrez["HGNC_ID"]=sapply(strsplit(hgnc2entrez$HGNC_ID,":"),"[",2) 
panther_orthologs <- merge(panther_orthologs,hgnc2entrez,by.x = "human_gene_id",by.y = "HGNC_ID")

mgi2entrez <- select(read_delim("C:/Users/inbal/Google Drive/shen-orr/panther/MGI_EntrezGene.csv", delim = "\t", col_names = c("mgi_ID","_","_","_","_","_","_","_","mouse_entrez_ID","_","_","_","_","_","_")),c("mgi_ID","mouse_entrez_ID"))
mgi2entrez["mgi_ID"] = sapply(strsplit(mgi2entrez$mgi_ID,":"),"[",2)
mgi2entrez <- mgi2entrez[!is.na(mgi2entrez$mouse_entrez_ID),]
panther_orthologs <- merge(panther_orthologs,mgi2entrez,by.x = "mouse_gene_id",by.y = "mgi_ID")

panther_orthologs <- select(panther_orthologs, c("ortholog_type","mouse_entrez_ID","human_entrez_ID"))

#classify genes as 1->1 / 1->many / many->1 / many->many
mouse_genes = unique(panther_orthologs$mouse_entrez_ID)
human_genes = unique(panther_orthologs$human_entrez_ID)

one2many_mouse = c()
many2many_mouse = c() 
many2one_mouse = c()
one2many_human = c()
many2many_human = c() 
many2one_human = c()
one2one=0
while(length(mouse_genes)>0) {
  mouse_gene = mouse_genes[1]
  mouse_orthologs=c(mouse_gene)
  human_orthologs=c()
  added=TRUE
  while(added==TRUE){
    mouse_orthologs_temp = unique(c(mouse_orthologs, (panther_orthologs %>% filter(human_entrez_ID %in% human_orthologs))$mouse_entrez_ID))
    human_orthologs_temp = unique(c(human_orthologs, (panther_orthologs %>% filter(mouse_entrez_ID %in% mouse_orthologs))$human_entrez_ID))
    if((length(mouse_orthologs_temp) == length(mouse_orthologs)) & (length(human_orthologs_temp) == length(human_orthologs))) added=FALSE
    mouse_orthologs = mouse_orthologs_temp
    human_orthologs = human_orthologs_temp
  }
  
  if(length(mouse_orthologs)==1 & length(human_orthologs)==1){
    one2one=one2one+1
    panther_orthologs[(panther_orthologs$mouse_entrez_ID %in% mouse_orthologs) | (panther_orthologs$human_entrez_ID %in% human_orthologs),]$ortholog_type="1:1"
  }
  else if(length(mouse_orthologs)>1 & length(human_orthologs)>1){
    many2many_mouse = c(many2many_mouse,mouse_orthologs)
    many2many_human = c(many2many_human, human_orthologs)
    panther_orthologs[(panther_orthologs$mouse_entrez_ID %in% mouse_orthologs) | (panther_orthologs$human_entrez_ID %in% human_orthologs),]$ortholog_type="n:m"
  }
  else if(length(mouse_orthologs)>1 & length(human_orthologs)==1){
    many2one_mouse = c(many2one_mouse,mouse_orthologs)
    many2one_human = c(many2one_human,human_orthologs)
    panther_orthologs[(panther_orthologs$mouse_entrez_ID %in% mouse_orthologs) | (panther_orthologs$human_entrez_ID %in% human_orthologs),]$ortholog_type="n:1"
  }
  else if(length(mouse_orthologs)==1 & length(human_orthologs)>1){
    one2many_mouse = c(one2many_mouse,mouse_orthologs)
    one2many_human = c(one2many_human,human_orthologs)
    panther_orthologs[(panther_orthologs$mouse_entrez_ID %in% mouse_orthologs) | (panther_orthologs$human_entrez_ID %in% human_orthologs),]$ortholog_type="1:m"
  }
  mouse_genes = mouse_genes[!mouse_genes %in% mouse_orthologs]  
}

write.csv(panther_orthologs,"C:/Users/inbal/Google Drive/shen-orr/panther/panther_orthologs.csv")
```

